### DISCUSSION TITLE
Thoughts on nuking the `include` folder and merging headers with sources?

### DISCUSSION THREAD

**Discussion by Rossmaxx** [LMMS/lmms#7190] [Project Management]
I am wondering if it would make any benefit to merge the headers to the src folder.+ I remember @jsmtux saying it might help with his core seperation effort. Thoughts?

**Comment by Rossmaxx** [comment]
I googled this topic a bit and found that "core" headers which are used only by the main application and not shared between programs are usually kept within src, only the headers which might be used by another program (like in the case of libraries) are the only ones in the seperate include folder. Should we follow the same?

**Reply by jsmtux** [reply]
Moving headers around is actually the most important part of the changes I did for the separation prototype, but it is no easy feat at all.
Now that every important header is in the same folder, there are no restrictions over what uses which. The idea I followed for the separation was to move headers as close as possible to the sources. Taking for instance `AboutDialog.h`, the first one, it should be next to `AboutDialog.cpp`, nothing in `src/core` has any reason for using it, so why make it avaliable there? 
Probably the intention here is just to move the whole folder just to inside `src`, but the closer, the better 😅
I am happy to help on this matters, as they will reduce the diff with my branch, but would feel safer to wait until there is a 1.3 branch to make sure no obscure target or configuration gets inadvertently broken.

**Reply by Rossmaxx** [reply]
> it is no easy feat at all.

That's why i opened this discussion before starting any work in order to not waste any work.

I appreciate your comment. I won't be starting any work for some time (atleast till i get nods from the main devs)

**Comment by Rossmaxx** [comment]
@LMMS/developers I'm considering doing this. Need some scrutiny before i start.

**Reply by Veratil** [reply]
There's a lot of work that would go into redoing all the includes, and touches pretty much every file. I think that's why we've avoided doing it.

**Reply by Rossmaxx** [reply]
I don't really have issues with committing time and effort. It's just a matter of "will the change be welcomed?". Also, I will do everything in a single PR as breaking the changes into parts will only complicate things for me.

**Reply by messmerd** [reply]
I like the idea. It's good to separate public headers from private headers. And besides the benefit for core separation, it would also let people know which headers can be used in LMMS plugins for example.

It might be worth mentioning that we already have at least one source file which has a private header.. See `src/core/UpgradeExtendedNoteRange.cpp` whose header is located in the same directory.

I think I'd be fine with following that convention for private headers - basically just move them from `include/` to wherever in `src/` their corresponding cpp file is. Unless people have different ideas for how to structure things.

**Reply by michaelgregorius** [reply]
Doing this will likely create merge conflicts in many open pull request. This should be taken into account as well. There are already several pull requests with merge conflicts due to other reorganizations and I guess increasing the number of these PRs will also decrease the chance that the respective developers will fix them because it will just become too much effort.

Perhaps the PRs should be managed/pruned before doing this?

**Reply by zonkmachine** [reply]
I wouldn't do any more rearrangement stuff this large until at least after 1.3 has been released.

**Reply by Rossmaxx** [reply]
Okk i will wait a bit more then, atleast till some old PRs get merged.
